File: Algorithms/Sorting/MergeSort.cpp
#include<bits/stdc++.h>
using namespace std;

void PrintArray(int *array, int n) {
  for (int i = 0; i < n; ++i)
    printf("%d ", array[i]);
  printf("\n");
}

void Merger(int arr[], int lo, int  mi, int hi){
    int *temp = new int[hi-lo+1];//temporary merger array
    int i = lo, j = mi + 1;//i is for left-hand,j is for right-hand
    int k = 0;//k is for the temporary array
    while(i <= mi && j <=hi){
        if(arr[i] <= arr[j])
            temp[k++] = arr[i++];
        else
            temp[k++] = arr[j++];
    }
    //rest elements of left-half
    while(i <= mi)
        temp[k++] = arr[i++];
    //rest elements of right-half
    while(j <= hi)
        temp[k++] = arr[j++];
    //copy the mergered temporary array to the original array
    for(k = 0, i = lo; i <= hi; ++i, ++k)
        arr[i] = temp[k];

    delete []temp;
}
void MergeSortHelper(int arr[], int lo, int hi){
    int mid;
    if(lo < hi){
        mid = (lo + hi) >> 1;
        MergeSortHelper(arr, lo, mid);
        MergeSortHelper(arr, mid+1, hi);
        Merger(arr, lo, mid, hi);
    }
}
void MergeSort(int arr[], int arr_size){
    MergeSortHelper(arr, 0, arr_size-1);
}

int main() {
  int array[] = {94, 0, 50, 95, 333, 3, 54, 456, 1, 28};
  int n = sizeof(array)/sizeof(array[0]);

  printf("Before Merge Sort:\n");
  PrintArray(array, n);

  MergeSort(array, n);

  printf("\nAfter Merge Sort:\n");
  PrintArray(array, n);

  return (0);
}

File: Algorithms/Sorting/MergeSort.java
package sorting;

public class MergeSort
{
	public static void sort(int[] in) {
	    if(in.length < 2) return; //do not need to sort
	    
	    int mid = in.length/2;
	    int left[] = new int[mid];
	    int right[] = new int[in.length-mid];
	    
	    for(int i=0; i<mid; i++) { //copy left
	        left[i] = in[i];
	    }
	    for(int i=0; i<in.length-mid; i++){ //copy right
	        right[i] = in[mid+i];
	    }
	    sort(left);
	    sort(right);
	    merge(left, right, in);
	}

	private static void merge(int[] a, int[] b, int[] all) {
	    int i=0, j=0, k=0;
	    while(i<a.length && j<b.length) { //merge back
	        if(a[i] < b[j]){
	            all[k] = a[i];
	            i++;
	        }else{
	            all[k] = b[j];
	            j++;
	        }
	        k++;
	    }
	    while(i<a.length) { //left remaining
	        all[k++] = a[i++];
	    }
	    while(j<b.length) { //right remaining
	        all[k++] = b[j++];
	    }
	}

	public static void main(String[] args) {
	    int[] a = {2,3,6,4,9,22,12,1};
	    
	    sort(a);
	    
	    System.out.println("Sorted array is:");
	    for(int j=0; j<a.length; j++) {
	        System.out.print(a[j] + " ");
	    }   
	 }
}